packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js

```diff
@@ -0,0 +1,109 @@
+// A thick client for getting information about FinancialContractFactory events.
+
+class FinancialContractFactoryEventClient {
+  /**
+   * @notice Constructs new FinancialContractEventClient.
+   * @param {Object} logger Winston module used to send logs.
+   * @param {Object} financialContractFactoryAbi ExpiringMultiPartyCreator or PerpetualMultiPartyCreator truffle ABI object.
+   * @param {Object} web3 Web3 provider from truffle instance.
+   * @param {String} financialContractFactoryAddress Ethereum address of the factory contract deployed on the current network.
+   * @param {Integer} startingBlockNumber Offset block number to index events from.
+   * @param {Integer} endingBlockNumber Termination block number to index events until. If not defined runs to `latest`.
+   * @return None or throws an Error.
+   */
+  constructor(
+    logger,
+    financialContractFactoryAbi,
+    web3,
+    financialContractFactoryAddress,
+    startingBlockNumber = 0,
+    endingBlockNumber = null,
+    contractType = "Perpetual"
+    // Default to PerpetualMultiParty for now since the first intended user is the funding rate proposer bot
+  ) {
+    this.logger = logger;
+    this.web3 = web3;
+
+    // Factory contract
+    this.financialContractFactory = new this.web3.eth.Contract(
+      financialContractFactoryAbi,
+      financialContractFactoryAddress
+    );
+    this.financialContractFactoryAddress = financialContractFactoryAddress;
+
+    // Factory Contract Events data structure to enable synchronous retrieval of information.
+    this.createdContractEvents = [];
+
+    // First block number to begin searching for events after.
+    this.firstBlockToSearch = startingBlockNumber;
+
+    // Last block number to end the searching for events at.
+    this.lastBlockToSearchUntil = endingBlockNumber;
+    this.lastUpdateTimestamp = 0;
+
+    this.contractType = contractType;
+  }
+  // Delete all events within the client
+  async clearState() {
+    this.createdContractEvents = [];
+  }
+
+  getAllCreatedContractEvents() {
+    return this.createdContractEvents;
+  }
+
+  // Returns the last update timestamp.
+  getLastUpdateTime() {
+    return this.lastUpdateTimestamp;
+  }
+
+  async update() {
+    // The last block to search is either the value specified in the constructor (useful in serverless mode) or is the
```

Comment:
Will we use this this way? Does this mean we'll only get contracts that are deployed within the last block window? Or are we effectively going to be searching from block 0 in serverless mode?

Reply:
This is a great point, I'm thinking the generalized solution is to make the user of this client (i.e. the FundingRateProposer), just set `startingBlock = 0` when constructing this client.

Reply:
Yeah, I actually think this could have a cool double-use. In the monitor, we may want to know about _new_ EMPs. Here, we want to know about _all_ EMPs.